### PR TITLE
Bad blocks - potential fix

### PR DESCRIPTION
**Pull request by MarekM25**
## Changes

We've identified instances of invalid blocks on the Gnosis network. During our debugging session, we uncovered that we were building on an incorrect state root (grandparent) rather than the parent state root.

The `TrySetState` function in `BlockProducerBase` ensures that we have the required state, so within the context of block production, we don't need to go through a loop. However, without this change, and in the event of a loop (or "branch") in the blockchainProcessor, we might end up building on top of an incorrect state root. This is because InitBranch sets the state root, and it could lead to the creation of an invalid block. The loop essentially goes backward in an attempt to align with the main chain, so if the main chain undergoes changes during block improvement, we might encounter this issue.

The loop scenario can occur as follows:
1. FCU with payload attributes for block 1000A (block building process initiated).
2. FCU block 1000B.
3. Main chain changed. (1000B)
4. A new block improvement for block 1000A begins.
5. Building on an incorrect state root. (999A instead of 1000A)
6. Resulting in an invalid block.

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] Not yet

### CODE REVIEW

**Comment by asdacap** [comment]
Can you add a description on step 4. Main chain changed to which block? And on step 6, which state root is it building on top off?

**Comment by MarekM25** [comment]
> Can you add a description on step 4. Main chain changed to which block? And on step 6, which state root is it building on top off?

@asdacap check the unit tests, I hope it will be more clear

> Can you add a description on step 4. Main chain changed to which block? And on step 6, which state root is it building on top off?

Updated steps + you can check unit test to understand it.

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs [hunk 1]
```diff
@@ -665,12 +665,13 @@ private ProcessingBranch PrepareProcessingBranch(Block suggestedBlock, Processin
             // otherwise some nodes would be missing
             bool notFoundTheBranchingPointYet = !_blockTree.IsMainChain(branchingPoint.Hash!);
             bool notReachedTheReorgBoundary = branchingPoint.Number > (_blockTree.Head?.Header.Number ?? 0);
-            preMergeFinishBranchingCondition = (notFoundTheBranchingPointYet || notReachedTheReorgBoundary);
+            bool notInBlockProduction = !options.ContainsFlag(ProcessingOptions.ProducingBlock);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:668]: notInForceProcessing?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:668]: Can this code be now removed:

https://github.com/NethermindEth/nethermind/pull/6212/files#diff-d3815f34e7dae0c4c0af79174d6c0d34c4c9d5c9e7259c23a56379cea9f72d51R534-R538

Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs]: > notInForceProcessing?

thx, changed.

oh yeah. This code looks unnecessary now, but I am not going to touch it in this PR. Let's analyze it more

**Review by Felperez02 — COMMENTED** [review]
Quiero saber bien de esto

**Pull request merged**